website/docs/r/vapp_vm.html.markdown

```diff
@@ -191,6 +214,18 @@ example for usage details. **Deprecates**: `network_name`, `ip`, `vapp_network_n
 
   * `ip_allocation_mode=NONE` - **`ip`** field can be omitted or set to an empty string "". Empty string may be useful when doing HCL variable interpolation.
 
+<a id="override_template_disk"></a>
+## Override template disk
+
```

Comment:
1. A general comment about the block is needed here.
2. Also, this is a good place to add an explanation when *not* to use this block and use internal disk resource instead. and a link to the internal disk resource.

Reply:
updated

Reply:
> Changes are ignored on update.

This is an interesting point. Won't this cause false expectations? Imagine a user who doesn't know this, he/she updates the override_template_disk config and apply succeeds, but disk is not changed. Any thoughts on how to address this?

CC: @Didainius @dataclouder

Reply:
If I understood override_template_disk correctly, changes to override_template_disk config will force a re-create of a VM with the correct parameter. The situation update config and apply without changes does not happen.

For me, it was unusual that there is a "one time shot" in VM definition. Usually you write the terraform template code, and then you manage the created objects with this code.

Reply:
> > Changes are ignored on update.
> 
> This is an interesting point. Won't this cause false expectations? Imagine a user who doesn't know this, he/she updates the override_template_disk config and apply succeeds, but disk is not changed. Any thoughts on how to address this?
> 
> CC: @Didainius @dataclouder

I could in code make check and throw warning if `override_template_disk` is changed

Reply:
> If I understood override_template_disk correctly, changes to override_template_disk config will force a re-create of a VM with the correct parameter. The situation update config and apply without changes does not happen.
> 
> For me, it was unusual that there is a "one time shot" in VM definition. Usually you write the terraform template code, and then you manage the created objects with this code.

Yes we understand that. Also issue is here that in UI it's the same and users requested such functionality - increase template disk size on VM creation.

Reply:
Don't know how difficult it is, but in theory, you can import the data of the template disks after VM is created. 
Let me try to explain with more details:
My idea is `override_template_disk` have 2 roles. 
If "Create", the data is used to fill the dialog known from UI. 
But in the end of `resourceVcdVAppVmCreate` in `resource_vcd_vapp_vm.go` you can import the disks (all disks present are template disks, since other internal disks will be build later after resourceVcdVAppVmCreate finished).
Now you are able to update the disk via `resourceVcdVAppVmUpdate` if `override_template_disk` has been changed.

From a user point of view this is easy to understand:
- If I want to modify the boot disk that is provided by the catalog_item, I need to modify the VM (because you will get it automatically with the VM)
- If I want to manage additional internal disks, I need to use `vcd_vm_internal_disk`

Reply:
Hi @buec011, thank you for your ideas. We tried many ways and various designs. And the end we have such implementation which has lowest limitations and works. Easiest way would have internal disk as typeSet in VM resource, but we couldn't due various issues and limitations:
* typeSet doesn't work well with different types and attributes of types in plan stage. Messup a lot - result user don't understand what will be final result. For e.g. change one internal disk - it will show that will change all of them, though in reality code will change only one.
* typeSet - using typeSet and disks isn't possible to implement normal update behaviour as you need keep track of diskID which is computed field and will be lost when any change is done in 
HCL as typeSet haschode will change and doesn't match in existing state file.
Use TypeList also isn't possible due that typeList is for ordered list and we don't have that.
As result separate resource allows us to implement the biggest part functionality and with little caveat.